In [1]:
import pandas as pd
import numpy as np

In [2]:
df_books=pd.read_csv("datos/good_reads_final.csv")
df_books.head()

author_average_rating author_gender                          author_genres  \
0                   4.01        female                    historical-fiction,   
1                   4.15          male  literature-fiction,mystery-thrillers,   
2                   4.00        female                               romance,   
3                   3.88          male                        fiction,memoir,   
4                   4.10        female                   young-adult,fantasy,   

   author_id            author_name                           author_page_url  \
0      74489    Victoria Thompson\n      /author/show/74489.Victoria_Thompson   
1     706255        Stieg Larsson\n         /author/show/706255.Stieg_Larsson   
2    5618190  Mimi Jean Pamfiloff\n  /author/show/5618190.Mimi_Jean_Pamfiloff   
3      37871          José Donoso\n             /author/show/37871.Jos_Donoso   
4      36122    Patricia C. Wrede\n       /author/show/36122.Patricia_C_Wrede   

   author_rating_count  author_review_count            birthplace  \
0                74399                 6268   United States\n       
1              3726435               142704          Sweden\n       
2                76496                 7975     United States\n     
3                 5522                  489           Chile\n       
4               291013                13453   United States\n       

   book_average_rating                                       book_fullurl  \
0                 4.02  https://www.goodreads.com/book/show/686717.Mur...   
1                 4.13  https://www.goodreads.com/book/show/2429135.Th...   
2                 3.99  https://www.goodreads.com/book/show/27833684-t...   
3                 4.14  https://www.goodreads.com/book/show/382975.The...   
4                 4.01  https://www.goodreads.com/book/show/64207.Sorc...   

    book_id                                         book_title  genre_1  \
0    686717               \n      Murder on St. Mark's Place\n  Mystery   
1   2429135          \n      The Girl with the Dragon Tattoo\n  Fiction   
2  27833684                     \n      Tailored for Trouble\n  Romance   
3    382975                \n      The Obscene Bird of Night\n  Fiction   
4     64207  \n      Sorcery & Cecelia: or The Enchanted Ch...  Fantasy   

           genre_2  num_ratings  num_reviews pages     publish_date  score  
0       Historical         5260          375   277             2000   3230  
1          Mystery      2229163        65227   465      August 2005   3062  
2     Contemporary         2151          391   354             2016   4585  
3  Magical Realism         1844          173   438             1970   1533  
4      Young Adult        17051         1890   326  April 15th 1988   2105

¿Cuántas filas y columnas tenemos?

In [3]:
df_books.shape

(22891, 20)

Vamos a echar primero un vistazo a las columnas que tenemos

In [4]:
df_books.columns

Index(['author_average_rating', 'author_gender', 'author_genres', 'author_id',
       'author_name', 'author_page_url', 'author_rating_count',
       'author_review_count', 'birthplace', 'book_average_rating',
       'book_fullurl', 'book_id', 'book_title', 'genre_1', 'genre_2',
       'num_ratings', 'num_reviews', 'pages', 'publish_date', 'score'],
      dtype='object')

No aparecen espacios en los nombres, pero vamos a poner los nombres de las columnas con la primera letra en mayúscula.

Usamos una list comprehension.


In [5]:
df_books.columns = [i.capitalize() for i in df_books.columns]
df_books.head(2)

Author_average_rating Author_gender                          Author_genres  \
0                   4.01        female                    historical-fiction,   
1                   4.15          male  literature-fiction,mystery-thrillers,   

   Author_id          Author_name                       Author_page_url  \
0      74489  Victoria Thompson\n  /author/show/74489.Victoria_Thompson   
1     706255      Stieg Larsson\n     /author/show/706255.Stieg_Larsson   

   Author_rating_count  Author_review_count            Birthplace  \
0                74399                 6268   United States\n       
1              3726435               142704          Sweden\n       

   Book_average_rating                                       Book_fullurl  \
0                 4.02  https://www.goodreads.com/book/show/686717.Mur...   
1                 4.13  https://www.goodreads.com/book/show/2429135.Th...   

   Book_id                                 Book_title  Genre_1     Genre_2  \
0   686717       \n      Murder on St. Mark's Place\n  Mystery  Historical   
1  2429135  \n      The Girl with the Dragon Tattoo\n  Fiction     Mystery   

   Num_ratings  Num_reviews Pages Publish_date  Score  
0         5260          375   277         2000   3230  
1      2229163        65227   465  August 2005   3062

Hay varias columnas que no nos interesa mantener, por lo que vamos a crear un nuevo DataFrame con las columnas con las que vamos a trabajar.

Vamos a hacerlo con el método drop. Podemos utilizar el metodo inplace=True o bien asignarle otra variable, que es lo que voya  hacer en este caso.

In [6]:
eliminar=["Author_id","Author_page_url","Book_fullurl", "Book_id"]
df_books2=df_books.drop(eliminar, axis = 1)
df_books2.head(2)

Author_average_rating Author_gender                          Author_genres  \
0                   4.01        female                    historical-fiction,   
1                   4.15          male  literature-fiction,mystery-thrillers,   

           Author_name  Author_rating_count  Author_review_count  \
0  Victoria Thompson\n                74399                 6268   
1      Stieg Larsson\n              3726435               142704   

             Birthplace  Book_average_rating  \
0   United States\n                     4.02   
1          Sweden\n                     4.13   

                                  Book_title  Genre_1     Genre_2  \
0       \n      Murder on St. Mark's Place\n  Mystery  Historical   
1  \n      The Girl with the Dragon Tattoo\n  Fiction     Mystery   

   Num_ratings  Num_reviews Pages Publish_date  Score  
0         5260          375   277         2000   3230  
1      2229163        65227   465  August 2005   3062

Author gender utiliza dos tipos de separadores: "," y "-". Vamos a unificarlo todo con "-"

In [7]:
def sustituir(genero):
    return genero.replace("," , "-")

df_books2["Author_genres"]=df_books2["Author_genres"].apply(sustituir)
df_books2.head(2)

Author_average_rating Author_gender                          Author_genres  \
0                   4.01        female                    historical-fiction-   
1                   4.15          male  literature-fiction-mystery-thrillers-   

           Author_name  Author_rating_count  Author_review_count  \
0  Victoria Thompson\n                74399                 6268   
1      Stieg Larsson\n              3726435               142704   

             Birthplace  Book_average_rating  \
0   United States\n                     4.02   
1          Sweden\n                     4.13   

                                  Book_title  Genre_1     Genre_2  \
0       \n      Murder on St. Mark's Place\n  Mystery  Historical   
1  \n      The Girl with the Dragon Tattoo\n  Fiction     Mystery   

   Num_ratings  Num_reviews Pages Publish_date  Score  
0         5260          375   277         2000   3230  
1      2229163        65227   465  August 2005   3062

De las columnas Author_name, Birthplace y Book_title vamos a eliminar las "\n"

In [8]:
def eliminar(n):
    return n.replace("\n","")

df_books2["Author_name"]=df_books2["Author_name"].apply(eliminar)
df_books2["Book_title"]=df_books2["Book_title"].apply(eliminar)
df_books2["Birthplace"]=df_books2["Birthplace"].apply(eliminar)
df_books2.head(2)

Author_average_rating Author_gender                          Author_genres  \
0                   4.01        female                    historical-fiction-   
1                   4.15          male  literature-fiction-mystery-thrillers-   

         Author_name  Author_rating_count  Author_review_count  \
0  Victoria Thompson                74399                 6268   
1      Stieg Larsson              3726435               142704   

           Birthplace  Book_average_rating  \
0   United States                     4.02   
1          Sweden                     4.13   

                              Book_title  Genre_1     Genre_2  Num_ratings  \
0             Murder on St. Mark's Place  Mystery  Historical         5260   
1        The Girl with the Dragon Tattoo  Fiction     Mystery      2229163   

   Num_reviews Pages Publish_date  Score  
0          375   277         2000   3230  
1        65227   465  August 2005   3062

La columna de fecha parece que tiene diferentes maneras de mostrarla. Vamos a verlo:

In [9]:
df_books2["Publish_date"]

0                      2000
1               August 2005
2                      2016
3                      1970
4           April 15th 1988
                ...        
22886                  2015
22887                  2014
22888      January 1st 2014
22889    December 28th 2012
22890                  2014
Name: Publish_date, Length: 22891, dtype: object

Tenemos las fechas como:
- años 
- meses y años  
- mes, día y año

Quiero quedarme solo con los años. Vamos a utilizar Regex para hacerlo de una manera mucho más sencilla.

In [10]:
df_books2["Publish_date"] = df_books2["Publish_date"].str.extract(r'(\d{4})')
df_books2.head()

Author_average_rating Author_gender                          Author_genres  \
0                   4.01        female                    historical-fiction-   
1                   4.15          male  literature-fiction-mystery-thrillers-   
2                   4.00        female                               romance-   
3                   3.88          male                        fiction-memoir-   
4                   4.10        female                   young-adult-fantasy-   

           Author_name  Author_rating_count  Author_review_count  \
0    Victoria Thompson                74399                 6268   
1        Stieg Larsson              3726435               142704   
2  Mimi Jean Pamfiloff                76496                 7975   
3          José Donoso                 5522                  489   
4    Patricia C. Wrede               291013                13453   

           Birthplace  Book_average_rating  \
0   United States                     4.02   
1          Sweden                     4.13   
2     United States                   3.99   
3           Chile                     4.14   
4   United States                     4.01   

                                          Book_title  Genre_1  \
0                         Murder on St. Mark's Place  Mystery   
1                    The Girl with the Dragon Tattoo  Fiction   
2                               Tailored for Trouble  Romance   
3                          The Obscene Bird of Night  Fiction   
4        Sorcery & Cecelia: or The Enchanted Choc...  Fantasy   

           Genre_2  Num_ratings  Num_reviews Pages Publish_date  Score  
0       Historical         5260          375   277         2000   3230  
1          Mystery      2229163        65227   465         2005   3062  
2     Contemporary         2151          391   354         2016   4585  
3  Magical Realism         1844          173   438         1970   1533  
4      Young Adult        17051         1890   326         1988   2105

Ahora que tenemos la información de la manera que queremos, vamos a empezar a analizarla más a fondo.
Vamos a ver el numero de Nan que tenemos y que porcentaje representan.

In [11]:
nulos_numero=pd.DataFrame(df_books2.isnull().sum())
print(nulos_numero[nulos_numero[0]>0])

nulos_porcentaje = pd.DataFrame(df_books2.isnull().sum()/df_books2.shape[0]*100)
print(nulos_porcentaje[nulos_porcentaje[0] > 0])

                0
Publish_date  813
                     0
Publish_date  3.551614


Tenemos 813 valores NaN que representan un 3,56% del total de datos en Publish_date.

Este dato habrá que tenerlo en cuenta cuando querramos hacer relaciones por año. Al ser fechas de publicación, no creo adecuado el usar la moda de la columna. Creo que en este caso, y al representar tan solo un 3,56% sería preferible ignorar esas filas en los cálculos.

Pasamos a ver si hay algun duplicado:

In [12]:
df_books2.duplicated().sum()

0

No hay duplicados.

Vamos a ver los principlaes estadísticos de las variables numéricas y categóricas

In [13]:
df_books2.describe().T

count           mean            std    min      25%  \
Author_average_rating  22891.0       3.960368       0.240421   1.82     3.81   
Author_rating_count    22891.0  172031.915687  654690.184879   6.00  4324.50   
Author_review_count    22891.0    9369.795640   24949.832609   0.00   545.00   
Book_average_rating    22891.0       3.951456       0.291317   0.00     3.77   
Num_ratings            22891.0   46683.487790  180069.814253   0.00   820.00   
Num_reviews            22891.0    2324.754794    6837.503574   0.00   106.00   
Score                  22891.0    3893.028483   11022.018268  55.00   832.00   

                            50%        75%         max  
Author_average_rating      3.97       4.12         5.0  
Author_rating_count    24635.00  111337.00  21117318.0  
Author_review_count     2273.00    8262.50    516745.0  
Book_average_rating        3.96       4.14         5.0  
Num_ratings             4403.00   20143.00   3820921.0  
Num_reviews              384.00    1504.00    147696.0  
Score                   1727.00    3597.50    598270.0

Vemos que la desviación estándar en la mayoría de los casos es enorme debido a la disparidad entre min y max.

In [14]:
df_books2.describe(include = "object").T

count unique              top   freq
Author_gender  22891      2             male  12201
Author_genres  22891   1776         romance-   2595
Author_name    22891  12156       Sara Gruen      6
Birthplace     22891    435                    4008
Book_title     22891  16350        Forbidden     16
Genre_1        22891    124          Romance   3397
Genre_2        22891    157          Romance   3042
Pages          22891    964              320    506
Publish_date   22078    306             2013   2688

Hay 124 géneros de libro. Vamos a ver si alguno que se pueda agrupar para poder simplificar la lista.

In [15]:
genre1=df_books2["Genre_1"].unique()
genre1

array(['Mystery', 'Fiction', 'Romance', 'Fantasy', 'Sequential Art',
       'Young Adult', 'Nonfiction', 'Paranormal', 'Historical',
       'Science Fiction', 'Horror', 'Classics', 'Autobiography',
       'Womens Fiction', 'New Adult', 'Contemporary', 'Humor', 'History',
       'Childrens', 'Lgbt', 'Sports and Games', 'War', 'Dark',
       'Philosophy', 'Poetry', 'Christian', 'Adventure', 'Anthologies',
       'Christian Fiction', 'Warfare', 'Erotica', 'Biography',
       'Adult Fiction', 'Short Stories', 'Animals', 'Science',
       'Shapeshifters', 'Self Help', 'Politics', 'Music', 'Religion',
       'Plays', 'Mental Health', 'Thriller', 'Crime', 'Mythology',
       'Travel', 'Military History', 'Language', 'Literature',
       'Psychology', 'Novels', 'Feminism', 'Economics', 'Business',
       'Media Tie In', 'World War II', 'Cultural',
       'Science Fiction Fantasy', 'Death', 'Lds', 'Northern Africa',
       'Design', 'Magical Realism', 'Apocalyptic', 'Realistic Fiction',
       

Genre_1 y Genre_2 tenemos 124 y 157 valores únicos, vamos a unificarlos un poco a ver cuanto podemos bajar ese número.

In [16]:
def sustituir(genre):
    if "Art" in genre:
        return "Art" 
    elif "War" in genre:
        return "War"
    elif "Fiction" in genre:
        return "Fiction"
    elif "Comic" in genre:
        return "Comic"
    elif "Glbt" in genre:
        return "Lgbt"
    elif "Romance" in genre:
        return "Romance"
    elif "Fantasy" in genre:
        return "Fantasy"
    elif "Finish" in genre:
        return "Unfinished"
    elif "Thriller" in genre:
        return "Thriller"
    elif "Novella" in genre:
        return "Novels"
    elif "Literature" in genre:
        return "Literature"
    elif "Sports" in genre:
        return "Sports"   
    elif "Christian" in genre:
        return "Christianity"

    else:
        return genre
   


df_books2["Genre_1"]=df_books2["Genre_1"].apply(sustituir)
df_books2["Genre_2"]=df_books2["Genre_2"].apply(sustituir)
df_books2.describe(include = "object").T




count unique              top   freq
Author_gender  22891      2             male  12201
Author_genres  22891   1776         romance-   2595
Author_name    22891  12156       Sara Gruen      6
Birthplace     22891    435                    4008
Book_title     22891  16350        Forbidden     16
Genre_1        22891    100          Fiction   3532
Genre_2        22891    126          Fiction   4492
Pages          22891    964              320    506
Publish_date   22078    306             2013   2688

Hemos bajado a 101 y 126, aun demasiados.

Según el IBIC (https://www.culturaydeporte.gob.es/dam/jcr:160ed8fc-99e8-477d-89ff-0bb688aaecc5/clasificacionibic.pdf) los géneros literarios son solo 19, así que vamos a agrupar los subgéneros en esos 19:

In [17]:
genres={
    "Artes":["Art", 'Couture','Media Tie In','Cultural','Design'],
    "Biografía e historias reales":["Nonfiction", 'Autobiography',  'Biography','Own'], 
    "Lengua":["Language"],
    "Literatura y estudios literarios":["Classics", "Literature",'Writing','Reference', 'Poetry'],
    "Enseñanza de la lengua inglesa (ELI)":"", 
    "Ficción y temas afines":["Mystery",'Retellings',"Fiction","Romance","Fantasy", "Paranormal","Horror", 'Contemporary', "Adventure", "New Adult","Dark","Erotica",'Short Stories', 'Shapeshifters', 'Thriller', 'Crime', 'Dungeons and Dragons','Suspense', 'Apocalyptic', 'Magical Realism',  'Novels', 'Adult','Epic', 'Westerns','Drama', 'Fairy Tales',"Love" ],
    "Consulta, información y temas interdisciplinarios":"",
    "Humanidades (historia, arqueología, filosofía, religión)":["Historical",'Amish',  'Teaching', "Christianity","Anthologies","Mythology", 'Religion',"Military History","Philosophy","History","War","Lds",'Death'],
    "Sociedad y ciencias sociales":['Social Science','Sociology','Politics', 'Feminism'],
    "Economía, finanzas, empresa y gestión":['Leadership','Economics','Business'], 
    "Derecho":"",
    "Medicina":['Medical','Psychology'],
    "Matemáticas y ciencia":['Science'],
    "Ciencias de la tierra, geografía, medioambiente, planificación": ["United States","Eastern Africa",'Northern Africa', 'Animals'],
    "Tecnología, ingeniería, agricultura":"",
    "Computación e informática":"",
    "Salud y desarrollo personal":['How To', 'Health','Self Help', 'Mental Health'],
    "Estilo de vida, deporte y ocio":["Sexuality", 'Polyamorous','Spirituality',"Parenting","Gardening",'Football', "Marriage","Family","Inspirational","Holiday","Music","Travel","Plays","Humor","Lgbt", 'Food and Drink','Relationships','Occult', 'Esoterica', 'Sports'],
    "Infantiles, juveniles y didácticos":["Young Adult",'Pulp', 'Manga','Comics', 'Childrens',"Education"]
    }
sum(map(len,genres.values()))


99

In [18]:
df_books2.head()

def ibic (genero):
    for k,v in genres.items():
        if genero in v:
            return k
       

df_books2["Main_genre"]=df_books2["Genre_1"].apply(ibic)

df_books2.head(3)


Author_average_rating Author_gender                          Author_genres  \
0                   4.01        female                    historical-fiction-   
1                   4.15          male  literature-fiction-mystery-thrillers-   
2                   4.00        female                               romance-   

           Author_name  Author_rating_count  Author_review_count  \
0    Victoria Thompson                74399                 6268   
1        Stieg Larsson              3726435               142704   
2  Mimi Jean Pamfiloff                76496                 7975   

           Birthplace  Book_average_rating  \
0   United States                     4.02   
1          Sweden                     4.13   
2     United States                   3.99   

                              Book_title  Genre_1       Genre_2  Num_ratings  \
0             Murder on St. Mark's Place  Mystery    Historical         5260   
1        The Girl with the Dragon Tattoo  Fiction       Mystery      2229163   
2                   Tailored for Trouble  Romance  Contemporary         2151   

   Num_reviews Pages Publish_date  Score              Main_genre  
0          375   277         2000   3230  Ficción y temas afines  
1        65227   465         2005   3062  Ficción y temas afines  
2          391   354         2016   4585  Ficción y temas afines

Vamos a crear un subframe con las columnas que más nos interesan de momento:

In [19]:
df_generos=df_books2[["Author_gender","Main_genre","Genre_1","Genre_2"]]
df_generos.describe(include = "object").T

count unique                     top   freq
Author_gender  22891      2                    male  12201
Main_genre     22867     14  Ficción y temas afines  13462
Genre_1        22891    100                 Fiction   3532
Genre_2        22891    126                 Fiction   4492

In [20]:
df_generos["Main_genre"].value_counts()

Ficción y temas afines                                            13462
Infantiles, juveniles y didácticos                                 2940
Humanidades (historia, arqueología, filosofía, religión)           2256
Biografía e historias reales                                       1255
Literatura y estudios literarios                                   1019
Artes                                                               725
Estilo de vida, deporte y ocio                                      683
Sociedad y ciencias sociales                                        111
Economía, finanzas, empresa y gestión                               107
Salud y desarrollo personal                                         103
Matemáticas y ciencia                                                74
Medicina                                                             69
Ciencias de la tierra, geografía, medioambiente, planificación       48
Lengua                                                          

Vemos que ficción tiene muchas más entradas que ningún otro género. Vamos a ver cuales serían los subgéneros más populares dentro de Ficción y ver si ha sid escrito por mujeres u hombres.

Para ellos vamos primero a extraerlos.

In [21]:
ficcion=df_generos.loc[df_generos['Main_genre'] == 'Ficción y temas afines']
ficcion

ficcion.groupby(['Main_genre']).value_counts()


Main_genre              Author_gender  Genre_1      Genre_2      
Ficción y temas afines  female         Romance      Romance          841
                                       Fantasy      Young Adult      707
                        male           Fiction      Fiction          568
                                       Fantasy      Young Adult      458
                                                    Fiction          380
                                                                    ... 
                                       Fairy Tales  Fantasy            1
                                       Fantasy      Action             1
                                                    Apocalyptic        1
                                                    Business           1
                                       Westerns     Short Stories      1
Length: 547, dtype: int64

Hay veces que los títulos cortos y largos se ponen de moda en diferentes años a raíz de un Bestseller, vamos a ver si hay algun patrón de ello, vamos a usar los años desde el 2000.

In [22]:
def palabras (titulo):
    if len(titulo.split())<=2:
        return "Corto"
    else: 
        return "Largo"

df_books2["Len_título"]=df_books2["Book_title"].apply(palabras)

df_books2["Len_años"]=""

df_len_años=df_books2.groupby(['Publish_date','Len_título'])["Len_años"].count().reset_index()

df_len_años['Publish_date'] = df_len_años['Publish_date'].astype("int64")
df_len_años.loc[(df_len_años['Publish_date'] >= 2000)]

Publish_date Len_título  Len_años
424          2000      Corto        63
425          2000      Largo       156
426          2001      Corto        70
427          2001      Largo       176
428          2002      Corto        68
429          2002      Largo       198
430          2003      Corto        93
431          2003      Largo       225
432          2004      Corto        93
433          2004      Largo       250
434          2005      Corto        98
435          2005      Largo       279
436          2006      Corto       158
437          2006      Largo       354
438          2007      Corto       174
439          2007      Largo       344
440          2008      Corto       249
441          2008      Largo       440
442          2009      Corto       308
443          2009      Largo       477
444          2010      Corto       472
445          2010      Largo       577
446          2011      Corto       751
447          2011      Largo       846
448          2012      Corto      1041
449          2012      Largo      1101
450          2013      Corto      1240
451          2013      Largo      1448
452          2014      Corto       851
453          2014      Largo      1082
454          2015      Corto       448
455          2015      Largo       674
456          2016      Corto       276
457          2016      Largo       501
458          2017      Corto       142
459          2017      Largo       322
460          2018      Corto       213
461          2018      Largo       399
462          2019      Corto        18
463          2019      Largo        24